### DISCUSSION TITLE
Trilium JSON Editor

### DISCUSSION THREAD

**Discussion by sottey** [TriliumNext/Trilium#4293] [Show and tell]
Check out a JSON editor for your notes in Trilium!

<img width="903" alt="image" src="https://github.com/zadam/trilium/assets/8494620/927bd97a-2566-4ce9-a383-3fe5d6cff795">

You can format cells, add rows and columns, use formulas, autosave changes, and more.

I have written this over the past few weeks, and used it daily for habit tracking, budget tracking, etc. I have found it very useful so I am releasing it for everyone.

This could not have been done without the amazing work done on the [x-spreadsheet project](https://github.com/myliang/x-spreadsheet). Be sure to check them out.

This editor gives you a table based editor for single level JSON notes in Trilium. Read more about it in the README.

[Trilium JSON Editor v0.1](https://github.com/sottey/tje)

**Comment by justyns** [comment]
This is pretty awesome, thanks for sharing!   

I have a few lists like books/tv shows to read/watch that I keep in a kanban-style board currently and was considering how to make something similar in Trilium.  I like this idea better though, it's a lot better imo than having to create a new note for every book/show/auther/etc if I'm not going to have any content other than metadata.

One feature request:  Can you add a link to open the json note directly?   And/or just show the full path of the note instead of the title.   There are a couple notes showing up in the dropdown that I have no idea where they're coming from or have duplicate titles

**Reply by sottey** [reply]
For sure! Great idea!

**Reply by sottey** [reply]
I have added that functionality. It isn't the most graceful code, but I wanted to get it out for you. Hopefully someone has a better approach. I basically get the "best note path", replace the commas with "/" and make that the link. It works but, like I said, isn't the prettiest.

**Reply by zerebos** [reply]
Could you make use of `createNoteLink` from the frontend api? https://zadam.github.io/trilium/frontend_api/FrontendScriptApi.html with `parentToBranch` from the note?

(or createLink for 0.61+)

Edit: I also just realized `getBestNotePathString()` and `getBestNotePath()` both exist for frontend notes anyways so you don't have to use `runOnBackend` necessarily

**Reply by justyns** [reply]
I just tried the 0.2 version and the link works for me, it's definitely good enough for what I was wanting, thanks!

**Comment by zadam** [comment]
Very nice! I do have plans to introduce a similar looking spreadsheet functionality, but so far haven't discovered a suitable library for that.

Because these things are not easily switchable, I have some expectations of long term maintenance. Ideally it would be something like handsontable, but they are no longer open source. One promising library is https://github.com/ruilisi/fortune-sheet

**Reply by zerebos** [reply]
I was just toying around with fortune-sheet in a render note the other day, very cool library.

**Comment by ShivaFang** [comment]
Anyone get this to work?  I'm not sure what I've done wrong.  (I couldn't find a 'save import' option to uncheck, so that might be it)
![image](https://github.com/zadam/trilium/assets/8532805/27b9a61a-5038-4d12-9794-4483e9b3a17c)

**Reply by sottey** [reply]
If you right click on a node on the tree and select Import, a dialog box will appear with import options. 

Hope that helps!

**Reply by ShivaFang** [reply]
Better!  But it's not detecting JSON notes in the heirarchy.  (I also tried setting it to the top root node as instructed and that didn't work either)
I made sure to set all the note types as instructed on those files
![image](https://github.com/zadam/trilium/assets/8532805/a0b3a393-53d5-4edb-b4e0-4e7057b89ed5)

**Reply by ShivaFang** [reply]
I dug into your code and found the search string you use... and the search IS working...
![image](https://github.com/zadam/trilium/assets/8532805/7e3a074f-019f-4c57-8c11-175de36d88b9)

**Reply by ShivaFang** [reply]
I think I fixed it.  Turns out the notes were in the wrong hierarchy (I realised it when I realised the .js wasn't actually excecuting and I knew they had to be under tje)
I also had to change 'runonbackend' to runAsyncOnBackendWithManualTransactionHandling because Trillium threw an error.
Not sure why you are running that on the backend anyway.  api.searchForNotes() works just fine on the frontend.
![image](https://github.com/zadam/trilium/assets/8532805/2d9d442e-33cb-431d-ace1-125b0de1f242)
